### PR TITLE
DatastorePageable uses url-safe-string cursor for JSON serializing

### LINKED ISSUE #1769: Datastore Pageable/Page Cursor JSON serializing
The issue is that the Cursor in our DatastorePageable can't automatically be serialized into JSON.

Two things we discussed:

1. looks like the Cursor is just a wrapper for some bytes and can be created from bytes easily. So if we store bytes inside the Pageable instead of Cursor then Jackson might automatically handle it. 
2. just annotate the cursor property to not be included in the JSON. this means when the user gives back the Page it has no cursor, but it still has the offset number of rows, which we handle too. It should just be a bit less efficient.

### PR DESCRIPTION
**Pull request by ChengyuanZhao**
fixes #1769 
stores url safe String cursor for json compatibility

### CODE REVIEW

File: spring-cloud-gcp-samples/spring-cloud-gcp-data-datastore-basic-sample/pom.xml [hunk 1]
```diff
@@ -1,4 +1,4 @@
-<?xml version="1.0" encoding="UTF-8"?>
+ <?xml version="1.0" encoding="UTF-8"?>
```
Comment (meltsufin) [inline spring-cloud-gcp-samples/spring-cloud-gcp-data-datastore-basic-sample/pom.xml]: ??

File: spring-cloud-gcp-samples/spring-cloud-gcp-data-datastore-basic-sample/src/test/java/com/example/DatastoreBookshelfExampleTests.java [hunk 2]
```diff
@@ -64,6 +77,17 @@ public static void checkToRun() {
...
 
+	@Test
+	public void testSerializedPage() {
+		Book book = new Book("Book1", "Author1", 2019);
+		book.id = 12345678L;
+		this.bookRepository.save(book);
+		Awaitility.await().atMost(15, TimeUnit.SECONDS)
+				.until(() -> this.bookRepository.count() == 1);
+		assertThat(sendRequest("/allbooksserialized", null, HttpMethod.GET))
+				.contains("content\":[{\"id\":12345678}],\"pageable\":");
```
Comment (meltsufin) [inline spring-cloud-gcp-samples/spring-cloud-gcp-data-datastore-basic-sample/src/test/java/com/example/DatastoreBookshelfExampleTests.java]: Can you also verify the cursor?

File: spring-cloud-gcp-data-datastore/src/main/java/org/springframework/cloud/gcp/data/datastore/repository/query/DatastorePageable.java [hunk 3]
```diff
@@ -22,33 +22,44 @@
...
 
 /**
+ * A pageable implementation for Cloud Datastore that uses the cursor for efficient reads.
+ *
  * @author Dmitry Solomakha
+ * @author Chengyuan Zhao
  */
 public class DatastorePageable extends PageRequest {
-	private final Cursor cursor;
+	private final String urlSafeCursor;
```
Comment (meltsufin) [inline spring-cloud-gcp-data-datastore/src/main/java/org/springframework/cloud/gcp/data/datastore/repository/query/DatastorePageable.java]: Maybe add `getUrlSafeCursor` method?

File: spring-cloud-gcp-samples/spring-cloud-gcp-data-datastore-basic-sample/pom.xml (hunk 1 above)
Comment (ChengyuanZhao) [reply spring-cloud-gcp-samples/spring-cloud-gcp-data-datastore-basic-sample/pom.xml]: I think this must've been the auto-formating

File: spring-cloud-gcp-samples/spring-cloud-gcp-data-datastore-basic-sample/src/test/java/com/example/DatastoreBookshelfExampleTests.java (hunk 2 above)
Comment (ChengyuanZhao) [reply spring-cloud-gcp-samples/spring-cloud-gcp-data-datastore-basic-sample/src/test/java/com/example/DatastoreBookshelfExampleTests.java]: Unfortunately the cursor isn't actually part of the response. 
While the `DatastorePageable` type is what gets serialized into JSON, what gets written is just the base `Pageable` and doesn't have the cursor.

File: spring-cloud-gcp-data-datastore/src/main/java/org/springframework/cloud/gcp/data/datastore/repository/query/DatastorePageable.java (hunk 3 above)
Comment (ChengyuanZhao) [reply spring-cloud-gcp-data-datastore/src/main/java/org/springframework/cloud/gcp/data/datastore/repository/query/DatastorePageable.java]: They can retrieve the urlSafeString from the Cursor actually. In fact we don't have the string aspect of it visible at all to the outside. I prefer to keep it only in terms of Cursor externally.

Comment (ChengyuanZhao) [reply spring-cloud-gcp-data-datastore/src/main/java/org/springframework/cloud/gcp/data/datastore/repository/query/DatastorePageable.java]: Actually let me add that getter. Perhaps a public getter will cause the serializer to pick it up. Stay tuned

File: spring-cloud-gcp-samples/spring-cloud-gcp-data-datastore-basic-sample/pom.xml (hunk 1 above)
Comment (ChengyuanZhao) [reply spring-cloud-gcp-samples/spring-cloud-gcp-data-datastore-basic-sample/pom.xml]: fixed :)

File: spring-cloud-gcp-data-datastore/src/main/java/org/springframework/cloud/gcp/data/datastore/repository/query/DatastorePageable.java (hunk 3 above)
Comment (ChengyuanZhao) [reply spring-cloud-gcp-data-datastore/src/main/java/org/springframework/cloud/gcp/data/datastore/repository/query/DatastorePageable.java]: fixed :)

File: spring-cloud-gcp-samples/spring-cloud-gcp-data-datastore-basic-sample/src/test/java/com/example/DatastoreBookshelfExampleTests.java (hunk 2 above)
Comment (ChengyuanZhao) [reply spring-cloud-gcp-samples/spring-cloud-gcp-data-datastore-basic-sample/src/test/java/com/example/DatastoreBookshelfExampleTests.java]: fixed :)

File: spring-cloud-gcp-samples/spring-cloud-gcp-data-datastore-basic-sample/src/test/java/com/example/DatastoreBookshelfExampleTests.java [hunk 4]
```diff
@@ -64,6 +77,18 @@ public static void checkToRun() {
...
+	@Test
+	public void testSerializedPage() {
+		Book book = new Book("Book1", "Author1", 2019);
+		book.id = 12345678L;
+		this.bookRepository.save(book);
+		Awaitility.await().atMost(15, TimeUnit.SECONDS)
+				.until(() -> this.bookRepository.count() == 1);
+		String responseBody = sendRequest("/allbooksserialized", null, HttpMethod.GET);
+		assertThat(responseBody).contains("content\":[{\"id\":12345678}],\"pageable\":");
+		assertThat(responseBody).contains("\"urlSafeCursor\":");
```
Comment (meltsufin) [inline spring-cloud-gcp-samples/spring-cloud-gcp-data-datastore-basic-sample/src/test/java/com/example/DatastoreBookshelfExampleTests.java]: Can you check that it has a value using regex?

Comment (ChengyuanZhao) [reply spring-cloud-gcp-samples/spring-cloud-gcp-data-datastore-basic-sample/src/test/java/com/example/DatastoreBookshelfExampleTests.java]: added

**Pull request merged**